pkg/filewatch/filewatch.go

```diff
@@ -0,0 +1,122 @@
+package filewatch
+
+import (
+	"fmt"
+	"strings"
+	"sync"
+	"sync/atomic"
+
+	"github.com/fsnotify/fsnotify"
+)
+
+type Watch interface {
+	Add(path string, callback func()) error
+
+	Run(done <-chan struct{}) error
+	IsRunning() bool
+}
+
+func New() Watch {
+	return &watch{
+		callbacks: make(map[string]func()),
+	}
+}
+
+type watch struct {
+	lock      sync.Mutex
+	callbacks map[string]func()
+	running   atomic.Bool
+}
+
+var _ Watch = &watch{}
+
+func (w *watch) Add(path string, callback func()) error {
+	if w.running.Load() {
```

Comment:
Checking twice?

Reply:
This is a performance optimization. The first check is fast, because it's without the lock, and the second check is to make sure that the value was not changed between the first check and taking the lock.

Reply:
Does it help much? How likely is it that `Add` is called when the watcher is running?

Reply:
I don't know how to answer if it helps "much". This is a usual pattern when writing lock-free code with atomic operations. But I can remove it, performance is not that important in this case.
I was just used to the pattern.

Reply:
> How likely is it that Add is called when the watcher is running?

IMHO, implementation of an exported function should not depend on likely usages. User of the function can do anything.

Reply:
By asking how likely it is I wanted to know how much of a performance benefit this will actually give us. My assumption was that the likelyhood of this being called while the watcher already runs is pretty low. But we can keep it if it's an common pattern.

Reply:
Removed.